# Guided Capstone Step 4 - Pre-processing and Training Data Development

### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   **Pre-processing and Training Data Development**

5.  Modeling
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

In this step of the guided capstone, you'll revisit some of the steps you completed to solve questions in step 4. Revisiting these steps will help you get additional practice completing this kind of work.

**<font color='teal'> Start by loading the necessary packages, as you've done in the previous steps. Print out your current working directory to confirm that you are in the correct project directory. </font>**

In [1]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline

In [18]:
cd Springboard/springboard/GuidedCapstone/data

C:\Users\Gabe\Springboard\springboard\GuidedCapstone\data


**<font color='teal'>  Load the csv file you created in step 3. Remember, it should be saved inside your data subfolder. Print the first five rows.</font>**

In [19]:
df = pd.read_csv('step3_output.csv')
df.head()

Name    state  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort   Alaska         3939           2500        250      1   
1  Eaglecrest Ski Area   Alaska         2600           1540       1200      0   
2     Hilltop Ski Area   Alaska         2090            294       1796      0   
3     Arizona Snowbowl  Arizona        11500           2300       9200      0   
4  Sunrise Park Resort  Arizona        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  ...  SkiableTerrain_ac  \
0        0.0          0          2     2  ...             1610.0   
1        0.0          0          0     0  ...              640.0   
2        0.0          0          0     0  ...               30.0   
3        0.0          1          0     2  ...              777.0   
4        0.0          0          1     2  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  cluster  
0          85.0              150.0           550.0        2  
1          53.0               90.0             0.0        1  
2          34.0              152.0            30.0        1  
3          89.0              122.0             0.0        2  
4          78.0              104.0            80.0        2  

[5 rows x 27 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `State`. Add the dummies back to the dataframe and remove the original column for `State`. </font>**

In [20]:
states = pd.get_dummies(df.state, prefix='State')
states.head()
df = pd.concat([df, states], axis=1)
df = df.drop('state', axis=1)

In [21]:
df.head()

Name  summit_elev  vertical_drop  base_elev  trams  \
0       Alyeska Resort         3939           2500        250      1   
1  Eaglecrest Ski Area         2600           1540       1200      0   
2     Hilltop Ski Area         2090            294       1796      0   
3     Arizona Snowbowl        11500           2300       9200      0   
4  Sunrise Park Resort        11100           1800       9200      0   

   fastEight  fastSixes  fastQuads  quad  triple  ...  State_Rhode Island  \
0        0.0          0          2     2       0  ...                   0   
1        0.0          0          0     0       0  ...                   0   
2        0.0          0          0     0       1  ...                   0   
3        0.0          1          0     2       2  ...                   0   
4        0.0          0          1     2       3  ...                   0   

   State_South Dakota  State_Tennessee  State_Utah  State_Vermont  \
0                   0                0           0              0   
1                   0                0           0              0   
2                   0                0           0              0   
3                   0                0           0              0   
4                   0                0           0              0   

   State_Virginia  State_Washington  State_West Virginia  State_Wisconsin  \
0               0                 0                    0                0   
1               0                 0                    0                0   
2               0                 0                    0                0   
3               0                 0                    0                0   
4               0                 0                    0                0   

   State_Wyoming  
0              0  
1              0  
2              0  
3              0  
4              0  

[5 rows x 61 columns]

In [25]:
df.index = df.Name
print(df.index)
df = df.drop('Name', axis=1)
df.head()

Index(['Alyeska Resort', 'Eaglecrest Ski Area', 'Hilltop Ski Area',
       'Arizona Snowbowl', 'Sunrise Park Resort',
       'Yosemite Ski & Snowboard Area', 'Bear Mountain', 'Bear Valley',
       'Boreal Mountain Resort', 'Dodge Ridge',
       ...
       'Whitecap Mountain', 'Wilmot Mountain', 'Grand Targhee Resort',
       'Hogadon Basin', 'Jackson Hole', 'Meadowlark Ski Lodge',
       'Sleeping Giant Ski Resort', 'Snow King Resort',
       'Snowy Range Ski & Recreation Area', 'White Pine Ski Area'],
      dtype='object', name='Name', length=330)


summit_elev  vertical_drop  base_elev  trams  fastEight  \
Name                                                                           
Alyeska Resort              3939           2500        250      1        0.0   
Eaglecrest Ski Area         2600           1540       1200      0        0.0   
Hilltop Ski Area            2090            294       1796      0        0.0   
Arizona Snowbowl           11500           2300       9200      0        0.0   
Sunrise Park Resort        11100           1800       9200      0        0.0   

                     fastSixes  fastQuads  quad  triple  double  ...  \
Name                                                             ...   
Alyeska Resort               0          2     2       0       0  ...   
Eaglecrest Ski Area          0          0     0       0       4  ...   
Hilltop Ski Area             0          0     0       1       0  ...   
Arizona Snowbowl             1          0     2       2       1  ...   
Sunrise Park Resort          0          1     2       3       1  ...   

                     State_Rhode Island  State_South Dakota  State_Tennessee  \
Name                                                                           
Alyeska Resort                        0                   0                0   
Eaglecrest Ski Area                   0                   0                0   
Hilltop Ski Area                      0                   0                0   
Arizona Snowbowl                      0                   0                0   
Sunrise Park Resort                   0                   0                0   

                     State_Utah  State_Vermont  State_Virginia  \
Name                                                             
Alyeska Resort                0              0               0   
Eaglecrest Ski Area           0              0               0   
Hilltop Ski Area              0              0               0   
Arizona Snowbowl              0              0               0   
Sunrise Park Resort           0              0               0   

                     State_Washington  State_West Virginia  State_Wisconsin  \
Name                                                                          
Alyeska Resort                      0                    0                0   
Eaglecrest Ski Area                 0                    0                0   
Hilltop Ski Area                    0                    0                0   
Arizona Snowbowl                    0                    0                0   
Sunrise Park Resort                 0                    0                0   

                     State_Wyoming  
Name                                
Alyeska Resort                   0  
Eaglecrest Ski Area              0  
Hilltop Ski Area                 0  
Arizona Snowbowl                 0  
Sunrise Park Resort              0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing, standardize the scale of the features of the dataframe except the name of the resort, which you don't need in the dataframe for modeling so it can be droppped here as well. You should  hold out your response variable(s) so that you can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as your response for scaling and modeling. Later, you will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave, them in the development dataframe. </font>**

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selectionm import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. You will start by using the adult weekend ticket price as your response variable for modeling.</font>**

In [62]:
from sklearn.model_selection import train_test_split
x = df.drop('AdultWeekend', axis=1)
y = df.AdultWeekend
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.25)

In [63]:
print(x_train)
print(x_test)
print(y_train)
print(y_test)

                        summit_elev  vertical_drop  base_elev  trams  \
Name                                                                   
Snowbasin                      9350           2900       6450      3   
Toggenburg Mountain            2000            700       1300      0   
Hidden Valley Ski Area         2566            310       2316      0   
Pebble Creek Ski Area          8560           2200       6360      0   
Gunstock                       2300           1400        900      0   
...                             ...            ...        ...    ...   
Cataloochee Ski Area           5400            740       4660      0   
Boyne Mountain Resort          1120            500        620      0   
Mountain High                  8200           1600       6600      0   
Song Mountain                  1940            700       1240      0   
Dodge Ridge                    8200           1600       6600      0   

                        fastEight  fastSixes  fastQuads  quad  

Here, you'll start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5 - Modeling

You've cleaned and prepared the datasets, so now it's time to get into the most exciting part of this process: modeling! In this exercise, you will build three different models and compare each model's performance. In the end, you will choose the best model for proving insights to Big Mountain management.  

1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with Training Data Set

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [64]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(x_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

In [65]:
y_pred = model.predict(x_test)
# print("x=%s, Predicted=%s" % (x_test[0], y_test[0]))

In [66]:
type(y_pred)
print(y_pred[0])
x_test.head(1)

71.49352071320055


summit_elev  vertical_drop  base_elev  trams  fastEight  \
Name                                                                     
Loon Mountain         3050           2100        950      1        0.0   

               fastSixes  fastQuads  quad  triple  double  ...  \
Name                                                       ...   
Loon Mountain          0          3     1       1       3  ...   

               State_Rhode Island  State_South Dakota  State_Tennessee  \
Name                                                                     
Loon Mountain                   0                   0                0   

               State_Utah  State_Vermont  State_Virginia  State_Washington  \
Name                                                                         
Loon Mountain           0              0               0                 0   

               State_West Virginia  State_Wisconsin  State_Wyoming  
Name                                                                
Loon Mountain                    0                0              0  

[1 rows x 59 columns]

In [73]:
final = x_test
final['adult_weekend_predicted'] = y_pred
final['adult_weekend'] = y_test
final.head()

C:\Users\Gabe\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Gabe\Anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


summit_elev  vertical_drop  base_elev  trams  \
Name                                                                       
Loon Mountain                      3050           2100        950      1   
Pajarito Mountain Ski Area        10441           1410       9031      0   
Grand Geneva                       1086            211        875      0   
Ski Santa Fe                      12075           1725      10350      0   
Holiday Mountain                   1550            400       1150      0   

                            fastEight  fastSixes  fastQuads  quad  triple  \
Name                                                                        
Loon Mountain                     0.0          0          3     1       1   
Pajarito Mountain Ski Area        0.0          0          0     1       1   
Grand Geneva                      0.0          0          0     0       0   
Ski Santa Fe                      0.0          0          0     1       2   
Holiday Mountain                  0.0          0          0     1       0   

                            double  ...  State_Tennessee  State_Utah  \
Name                                ...                                
Loon Mountain                    3  ...                0           0   
Pajarito Mountain Ski Area       3  ...                0           0   
Grand Geneva                     3  ...                0           0   
Ski Santa Fe                     2  ...                0           0   
Holiday Mountain                 1  ...                0           0   

                            State_Vermont  State_Virginia  State_Washington  \
Name                                                                          
Loon Mountain                           0               0                 0   
Pajarito Mountain Ski Area              0               0                 0   
Grand Geneva                            0               0                 0   
Ski Santa Fe                            0               0                 0   
Holiday Mountain                        0               0                 0   

                            State_West Virginia  State_Wisconsin  \
Name                                                               
Loon Mountain                                 0                0   
Pajarito Mountain Ski Area                    0                0   
Grand Geneva                                  0                1   
Ski Santa Fe                                  0                0   
Holiday Mountain                              0                0   

                            State_Wyoming  adult_weekend_predicted  \
Name                                                                 
Loon Mountain                           0                71.493521   
Pajarito Mountain Ski Area              0                52.273056   
Grand Geneva                            0                37.361577   
Ski Santa Fe                            0                76.487352   
Holiday Mountain                        0                37.287338   

                            adult_weekend  
Name                                       
Loon Mountain                    64.16681  
Pajarito Mountain Ski Area       49.00000  
Grand Geneva                     49.00000  
Ski Santa Fe                     80.00000  
Holiday Mountain                 42.00000  

[5 rows x 61 columns]

In [68]:
explained_variance_score(
    y_test,
    y_pred,
    sample_weight=None,
    multioutput='uniform_average',
)

0.8056419045464196

In [69]:
mean_absolute_error(
    y_test,
    y_pred,
    sample_weight=None,
    multioutput='uniform_average',
)

5.326901037355229

## Review Model Outcomes — Iterate over additional models as needed.

**<font color='teal'> Print the intercept value from the linear model. </font>**

In [72]:
print(lm.intercept_)

12.098013845798839


**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the `AdultWeekend` resulting value. Also, because you took the time to scale your x values in the training data, you can compare each of the coefficients for the features to determine the feature's importance. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 

Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what you are looking for is the magnitude of impact on your response variable.

In [91]:
coeff_df = pd.DataFrame(lm.coef_, x.columns, columns=['Coefficient'])  
coeff_df.sort_values(by='Coefficient', ascending=False)


Coefficient
State_Tennessee       1.984366e+01
State_North Carolina  1.261864e+01
State_Virginia        1.023380e+01
State_Maryland        9.960662e+00
State_West Virginia   9.562504e+00
State_Massachusetts   9.521628e+00
State_Maine           6.553200e+00
State_New Hampshire   6.432207e+00
State_Rhode Island    3.939317e+00
State_Pennsylvania    3.394294e+00
State_Connecticut     3.085270e+00
State_Vermont         3.022566e+00
State_New York        2.237954e+00
State_Illinois        1.749788e+00
State_Nevada          1.495704e+00
triple                1.368394e+00
quad                  1.198258e+00
State_New Jersey      1.172897e+00
double                8.705386e-01
fastQuads             8.441525e-01
AdultWeekday          8.301458e-01
State_Wisconsin       7.575350e-01
TerrainParks          2.862924e-01
daysOpenLastYear      1.888860e-02
NightSkiing_ac        8.238012e-03
base_elev             4.122533e-03
vertical_drop         3.743057e-03
averageSnowfall       1.835788e-03
State_Indiana         2.309264e-14
Snow Making_ac       -1.051611e-04
SkiableTerrain_ac    -1.477496e-04
summit_elev          -3.337886e-03
projectedDaysOpen    -6.114437e-03
Runs                 -1.168935e-02
yearsOpen            -1.997350e-02
surface              -1.367242e-01
trams                -2.128377e-01
total_chairs         -2.317400e-01
fastSixes            -2.331083e-01
LongestRun_mi        -4.242364e-01
cluster              -1.206213e+00
State_Michigan       -1.510121e+00
State_Alaska         -2.730806e+00
State_Minnesota      -2.794678e+00
State_Missouri       -3.333584e+00
fastEight            -3.930413e+00
State_Utah           -4.050581e+00
State_Iowa           -4.382578e+00
State_Washington     -4.990887e+00
State_Colorado       -5.969736e+00
State_Arizona        -6.308511e+00
State_Oregon         -7.079927e+00
State_Ohio           -7.964512e+00
State_California     -7.971424e+00
State_Idaho          -8.495540e+00
State_South Dakota   -8.597883e+00
State_Wyoming        -9.201571e+00
State_Montana        -9.890764e+00
State_New Mexico     -1.030853e+01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that you care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

#### Model 2

In [98]:
# Define new dataframe.
columns = df.columns
state_columns = [ x for x in columns if "State" in x ]
X2 = df.drop('AdultWeekend', axis=1)
X2 = X2.drop(state_columns, axis=1)
y = df.AdultWeekend
X2.head()

summit_elev  vertical_drop  base_elev  trams  fastEight  \
Name                                                                           
Alyeska Resort              3939           2500        250      1        0.0   
Eaglecrest Ski Area         2600           1540       1200      0        0.0   
Hilltop Ski Area            2090            294       1796      0        0.0   
Arizona Snowbowl           11500           2300       9200      0        0.0   
Sunrise Park Resort        11100           1800       9200      0        0.0   

                     fastSixes  fastQuads  quad  triple  double  ...  \
Name                                                             ...   
Alyeska Resort               0          2     2       0       0  ...   
Eaglecrest Ski Area          0          0     0       0       4  ...   
Hilltop Ski Area             0          0     0       1       0  ...   
Arizona Snowbowl             1          0     2       2       1  ...   
Sunrise Park Resort          0          1     2       3       1  ...   

                     LongestRun_mi  SkiableTerrain_ac  Snow Making_ac  \
Name                                                                    
Alyeska Resort                 1.0             1610.0           113.0   
Eaglecrest Ski Area            2.0              640.0            60.0   
Hilltop Ski Area               1.0               30.0            30.0   
Arizona Snowbowl               2.0              777.0           104.0   
Sunrise Park Resort            1.2              800.0            80.0   

                     daysOpenLastYear  yearsOpen  averageSnowfall  \
Name                                                                
Alyeska Resort                  150.0       60.0            669.0   
Eaglecrest Ski Area              45.0       44.0            350.0   
Hilltop Ski Area                150.0       36.0             69.0   
Arizona Snowbowl                122.0       81.0            260.0   
Sunrise Park Resort             115.0       49.0            250.0   

                     AdultWeekday  projectedDaysOpen  NightSkiing_ac  cluster  
Name                                                                           
Alyeska Resort               65.0              150.0           550.0        2  
Eaglecrest Ski Area          47.0               90.0             0.0        1  
Hilltop Ski Area             30.0              152.0            30.0        1  
Arizona Snowbowl             89.0              122.0             0.0        2  
Sunrise Park Resort          74.0              104.0            80.0        2  

[5 rows x 24 columns]

In [99]:
#Build training parameters.
X_train2, X_test2, y_train2, y_test2 = train_test_split(X2, y, test_size=0.25)

#And model.
model2 = lm.fit(X_train2, y_train2)
y_pred2 = model.predict(X_test2)

**<font color='teal'> When reviewing your new model coefficients, note that `summit_elev` is now in the number two spot. From a managerial perspective, this is also difficult to change and highly correlated with `base_elev` and `vertical_drop`.  This time rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

In [101]:
coeff_df2 = pd.DataFrame(lm.coef_, X2.columns, columns=['Coefficient'])  
coeff_df2.sort_values(by='Coefficient', ascending=False)

Coefficient
fastEight             1.057239
quad                  0.938686
AdultWeekday          0.839188
fastQuads             0.411852
total_chairs          0.365333
triple                0.163723
double                0.121002
TerrainParks          0.109898
daysOpenLastYear      0.087611
NightSkiing_ac        0.006143
summit_elev           0.002720
yearsOpen             0.001003
SkiableTerrain_ac    -0.000092
Snow Making_ac       -0.000493
vertical_drop        -0.001320
base_elev            -0.003201
averageSnowfall      -0.006287
Runs                 -0.029187
projectedDaysOpen    -0.066073
LongestRun_mi        -0.188771
fastSixes            -0.207178
cluster              -0.664952
surface              -0.723647
trams                -1.396343

#### Model 3

In [109]:
#Define dataframe.
X3 = X2.drop(['summit_elev', 'vertical_drop', 'base_elev'], axis=1)

#Build training parameters.
X_train3, X_test3, y_train3, y_test3 = train_test_split(X3, y, test_size=0.25)

#And model.
model3 = lm.fit(X_train3, y_train3)
y_pred3 = model.predict(X_test3)

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell. You will explain your selection during the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.8056 | 5.3269 |-|
| Model 2. | 0.8790 | 5.9963 |'state'|
| Model 3. | 0.8225 | 5.9646 |'state','summit_elev','base_elev'|

In [110]:
#Running the numbers for above table.
print(explained_variance_score(
    y_test,
    y_pred,
    sample_weight=None,
    multioutput='uniform_average',
))
print(explained_variance_score(
    y_test2,
    y_pred2,
    sample_weight=None,
    multioutput='uniform_average',
))
print(explained_variance_score(
    y_test3,
    y_pred3,
    sample_weight=None,
    multioutput='uniform_average',
))

0.8056419045464196
0.8790353861417506
0.8225427477414448


In [112]:
print(mean_absolute_error(
    y_test,
    y_pred,
    sample_weight=None,
    multioutput='uniform_average',
))
print(mean_absolute_error(
    y_test2,
    y_pred2,
    sample_weight=None,
    multioutput='uniform_average',
))
print(mean_absolute_error(
    y_test3,
    y_pred3,
    sample_weight=None,
    multioutput='uniform_average',
))

5.326901037355229
5.996251958290206
5.964579425423817


Model Selection: